akka-typed/src/test/java/akka/typed/javadsl/AdapterTest.java

```diff
@@ -0,0 +1,342 @@
+  }
+
+  static class Untyped1 extends akka.actor.AbstractActor {
+    @Override
+    public Receive createReceive() {
+      return receiveBuilder()
+        .matchEquals("ping", s -> getSender().tell("pong", getSelf()))
+        .match(ThrowIt.class, t -> {
+          throw t;
+        })
+        .build();
+    }
+  }
+
+  static class Typed1 {
+    private final akka.actor.ActorRef ref;
+    private final akka.actor.ActorRef probe;
+
+    private Typed1(akka.actor.ActorRef ref, akka.actor.ActorRef probe) {
+      this.ref = ref;
+      this.probe = probe;
+    }
+
+    static Behavior<String> create(akka.actor.ActorRef ref, akka.actor.ActorRef probe) {
+      Typed1 logic = new Typed1(ref, probe);
+      return stateful(
+          (ctx, msg) -> logic.onMessage(ctx, msg),
+          (ctx, sig) -> logic.onSignal(ctx, sig));
+    }
+
+    Behavior<String> onMessage(ActorContext<String> ctx, String msg) {
+      if (msg.equals("send")) {
+        akka.actor.ActorRef replyTo = Adapter.toUntyped(ctx.getSelf());
+        ref.tell("ping", replyTo);
+        return same();
+      } else if (msg.equals("pong")) {
+        probe.tell("ok", akka.actor.ActorRef.noSender());
+        return same();
+      } else if (msg.equals("actorOf")) {
+        akka.actor.ActorRef child = Adapter.actorOf(ctx, untyped1());
+        child.tell("ping", Adapter.toUntyped(ctx.getSelf()));
+        return same();
+      } else if (msg.equals("watch")) {
+        Adapter.watch(ctx, ref);
+        return same();
+      } else if (msg.equals("supervise-stop")) {
+        akka.actor.ActorRef child = Adapter.actorOf(ctx, untyped1());
+        Adapter.watch(ctx, child);
+        child.tell(new ThrowIt3(), Adapter.toUntyped(ctx.getSelf()));
+        child.tell("ping", Adapter.toUntyped(ctx.getSelf()));
+        return same();
+      } else if (msg.equals("stop-child")) {
+        akka.actor.ActorRef child = Adapter.actorOf(ctx, untyped1());
+        Adapter.watch(ctx, child);
+        Adapter.stop(ctx, child);
+        return same();
+      } else {
+        return unhandled();
+      }
+    }
+
+    Behavior<String> onSignal(ActorContext<String> ctx, Signal sig) {
+      if (sig instanceof Terminated) {
+        probe.tell("terminated", akka.actor.ActorRef.noSender());
+        return same();
+      } else {
+        return unhandled();
+      }
+    }
+  }
+
+  static interface Typed2Msg {};
+  static final class Ping implements Typed2Msg {
+    public final ActorRef<String> replyTo;
+
+    public Ping(ActorRef<String> replyTo) {
+      this.replyTo = replyTo;
+    }
+  }
+  static final class StopIt implements Typed2Msg {}
+  static abstract class ThrowIt extends RuntimeException implements Typed2Msg {}
+  static class ThrowIt1 extends ThrowIt {}
+  static class ThrowIt2 extends ThrowIt {}
+  static class ThrowIt3 extends ThrowIt {}
+
+  static akka.actor.Props untyped2(ActorRef<Ping> ref, akka.actor.ActorRef probe) {
```

Comment:
To avoid `.narrow()` when calling this, this could be `ActorRef<? super Ping> ref` (yay use site variance). But then `? super Ping` would need to trickle down to the constructor and parameters of `Untyped2`. So `narrow()` seems to be less verbose and infectious.